#  Proyecto 2: <font color=blue>*Pandas Project*</font>
<img src=https://animals.sandiegozoo.org/sites/default/files/styles/image_grid_full_width/public/2017-04/%20leopard_shark_01.jpg>

# Conociendo los datos
Tareas:
- Crear df de problemas con columnas de: Tipo de dato, numero de nulos, problema principal, solucionado S/N
FALTA COLUMNA lOCATION, COUNTRY, AREA Y INVESTIGADOR 

## Importamos bibliotecas

In [159]:
import pandas as pd
import numpy as np
import re

## Abrimos el archivo 

In [245]:
df=pd.read_csv("data/GSAF5.csv",encoding="ISO-8859-1")

## Informacion inicial del archivo

### Observacion inicial del dataset (head, info, isna().sum())

In [246]:
df.head(10)

Case Number       Date  Year        Type        Country  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked            USA   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked            USA   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked            USA   
3    2016.09.17  17-Sep-16  2016  Unprovoked      AUSTRALIA   
4    2016.09.15  16-Sep-16  2016  Unprovoked      AUSTRALIA   
5  2016.09.15.R  15-Sep-16  2016        Boat      AUSTRALIA   
6    2016.09.11  11-Sep-16  2016  Unprovoked            USA   
7    2016.09.07  07-Sep-16  2016  Unprovoked            USA   
8    2016.09.06  06-Sep-16  2016  Unprovoked  NEW CALEDONIA   
9  2016.09.05.b  05-Sep-16  2016  Unprovoked            USA   

                Area                                         Location  \
0            Florida                 New Smyrna Beach, Volusia County   
1            Florida                 New Smyrna Beach, Volusia County   
2            Florida                 New Smyrna Beach, Volusia County   
3           Victoria                                 Thirteenth Beach   
4           Victoria                                      Bells Beach   
5  Western Australia                                          Bunbury   
6            Florida                    Ponte Vedra, St. Johns County   
7             Hawaii                                     Makaha, Oahu   
8     North Province                                           Koumac   
9     South Carolina  Kingston Plantation, Myrtle Beach, Horry County   

          Activity                    Name Sex   ...          Species   \
0          Surfing                    male    M  ...               NaN   
1          Surfing          Chucky Luciano    M  ...               NaN   
2          Surfing                    male    M  ...               NaN   
3          Surfing         Rory Angiolella    M  ...               NaN   
4          Surfing                    male    M  ...         2 m shark   
5          Fishing  Occupant: Ben Stratton  NaN  ...               NaN   
6           Wading                    male    M  ...    3' to 4' shark   
7         Swimming                  female    F  ...  Tiger shark, 10?   
8     Kite surfing            David Jewell    M  ...               NaN   
9  Boogie boarding          Rylie Williams    F  ...               NaN   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   
5   West Australian, 9/15/2016      2016.09.15.R-boat.pdf   
6          News4Jax, 9/11/2016  2016.09.11-PonteVedra.pdf   
7    Hawaii News Now, 9/7/2016        2016.09.07-Oahu.pdf   
8       TVANouvelles, 9/6/2016      2016.09.06-Jewell.pdf   
9            C. Creswell, GSAF  2016.09.05.b-Williams.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   
5  http://sharkattackfile.net/spreadsheets/pdf_di...   
6  http://sharkattackfile.net/spreadsheets/pdf_di...   
7  http://sharkattackfile.net/spreadsheets/pdf_di...   
8  http://sharkattackfile.net/spreadsheets/pdf_di...   
9  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Case Number             5992 non-null   object
 1   Date                    5992 non-null   object
 2   Year                    5992 non-null   int64 
 3   Type                    5992 non-null   object
 4   Country                 5949 non-null   object
 5   Area                    5590 non-null   object
 6   Location                5496 non-null   object
 7   Activity                5465 non-null   object
 8   Name                    5792 non-null   object
 9   Sex                     5425 non-null   object
 10  Age                     3311 non-null   object
 11  Injury                  5965 non-null   object
 12  Fatal (Y/N)             5973 non-null   object
 13  Time                    2779 non-null   object
 14  Species                 3058 non-null   object
 15  Inve

In [5]:
#Valores nulos
df.isna().sum()

Case Number                  0
Date                         0
Year                         0
Type                         0
Country                     43
Area                       402
Location                   496
Activity                   527
Name                       200
Sex                        567
Age                       2681
Injury                      27
Fatal (Y/N)                 19
Time                      3213
Species                   2934
Investigator or Source      15
pdf                          0
href formula                 1
href                         3
Case Number.1                0
Case Number.2                0
original order               0
Unnamed: 22               5991
Unnamed: 23               5990
dtype: int64

In [6]:
#Vamos a ver si hay algun duplicado en los valores
df.duplicated().value_counts()

False    5992
dtype: int64

In [7]:
df.shape

(5992, 24)

### Conclusiones iniciales tras Analisis inicial:


- Columnas con un mismo valor como case number (0,19,20) y original order: Comprobar diferencias y si son iguales borrar
- Se solapan las columnas Year y Date: Agruparla en una.
- En algunos casos los datos de la columna nombre se solapan con el sexo: Buscar solucion
- Las columnas pdf, href formula y href se solapan, intentar agrupar en una
- original order posible indice

#  Modificacion del dataset

## Localizacion de problemas y limpieza del dataset por columnas

FOTO DE LA TABLA INICIAL

### Columna: "original order","Case Number","Case Number.1","Case Number.2"

- La columna original order es un indice de casos en los que se repiten valores igual que las columnas Case Number
- Objetivo: Unica lista de numero de casos: lista con valores unicos y continuos
- Output: Nueva columna que usare indice


In [8]:
df[["original order","Case Number","Case Number.1","Case Number.2"]]

original order   Case Number Case Number.1 Case Number.2
0               5993  2016.09.18.c  2016.09.18.c  2016.09.18.c
1               5992  2016.09.18.b  2016.09.18.b  2016.09.18.b
2               5991  2016.09.18.a  2016.09.18.a  2016.09.18.a
3               5990    2016.09.17    2016.09.17    2016.09.17
4               5989    2016.09.15    2016.09.16    2016.09.15
...              ...           ...           ...           ...
5987               6       ND.0005       ND.0005       ND.0005
5988               5       ND.0004       ND.0004       ND.0004
5989               4       ND.0003       ND.0003       ND.0003
5990               3       ND.0002       ND.0002       ND.0002
5991               2       ND.0001       ND.0001       ND.0001

[5992 rows x 4 columns]

#### Analisis de original order

In [9]:
#ordeno la tabla por los valores original order
df.sort_values("original order",ascending=True,inplace=True)
df.head(3)

Case Number       Date  Year        Type             Country  \
5991     ND.0001  1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   
5990     ND.0002  1883-1889     0  Unprovoked              PANAMA   
5989     ND.0003  1900-1905     0  Unprovoked                 USA   

                  Area                             Location  Activity  \
5991  Eastern Province  Below the English fort, Trincomalee  Swimming   
5990               NaN                 Panama Bay 8ºN, 79ºW       NaN   
5989    North Carolina                       Ocracoke Inlet  Swimming   

                       Name Sex   ... Species   \
5991                   male    M  ...      NaN   
5990        Jules Patterson    M  ...      NaN   
5989  Coast Guard personnel    M  ...      NaN   

                    Investigator or Source                             pdf  \
5991                            S.W. Baker              ND-0001-Ceylon.pdf   
5990                   The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
5989  F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   

                                           href formula  \
5991  http://sharkattackfile.net/spreadsheets/pdf_di...   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
5991  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0001   
5990  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0002   
5989  http://sharkattackfile.net/spreadsheets/pdf_di...       ND.0003   

     Case Number.2 original order Unnamed: 22 Unnamed: 23  
5991       ND.0001              2         NaN         NaN  
5990       ND.0002              3         NaN         NaN  
5989       ND.0003              4         NaN         NaN  

[3 rows x 24 columns]

In [10]:
#Vamos a ver que valoers faltan
df["original order"].nunique()

5988

In [11]:
#Vamos a comprobar cuantos valores se repiten 
OE_ValoresRepetidos=len(df["original order"])-df["original order"].nunique()

In [12]:
#Comprobamos QUE valores se repiten
df["original order"].value_counts().head(OE_ValoresRepetidos)

5739    2
569     2
5661    2
3847    2
Name: original order, dtype: int64

In [13]:
#Comprobamos desde el 0 al numero de casos, que valores no estan, lopaso a index ya que me resuta mas facil
df2=df.set_index("original order")
for i in range(0,len(df2)):
    try:
        df2.loc[i]
    except KeyError:
        print(i)

0
1
571
3347
5724
5881


Problema de original order: 4 numeros repetidos y en el orden de casos se han saltado varios numeros
Opciones para corregirlo:
- Opcion A: Cambios en la misma columna:
        - del 2 al 568 correcto, pasamos el segundo valor 569 al 570 
        - del 571 al 3346 correcto, del3348 al primer valor 3847 hay que restarle 1
        - del 3848 al 5660 correcto, del segundo valor del 5661 a 5723 hay que sumarle 1
        - del 5725 a 5738 correcto, del segundo valor de 5739 al 5880 hay que sumarle 1
        - despues restamos 1 a todo y los casos empezaran desde 1 siendo valores unicos
- Opcion B:
        - ponerlo en orden y crear una nueva tabla con un rango de 1 a len

In [14]:
#cambio el indice con una lista de numeros desde el 1 hasta la longitud +1
lista=[i for i in range(1,len(df)+1)]


In [15]:
# añado la columna nueva y borro la antigua
df["Casos"]=lista
df.drop("original order",axis=1,inplace=True)

#### Analisis Case Number

In [16]:
#Cambio el indice a original a Casos 
df=df.set_index("Casos")

In [17]:
#Miramos cantidad de valores nulos
df[["Case Number","Case Number.1","Case Number.2"]].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5992 entries, 1 to 5992
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Case Number    5992 non-null   object
 1   Case Number.1  5992 non-null   object
 2   Case Number.2  5992 non-null   object
dtypes: object(3)
memory usage: 187.2+ KB


In [18]:
#Valores unicos que hay
df[["Case Number","Case Number.1","Case Number.2"]].nunique()

Case Number      5976
Case Number.1    5975
Case Number.2    5976
dtype: int64

In [19]:
#No hay valores nulos pero hay valores que se repiten y el formato de numero de caso varia
print(f'{len(df["Case Number"])-len(set(df["Case Number"]))} casos se repiten')

16 casos se repiten


In [20]:
#Vamos a ver que valores se repiten en Case Number y si son los mismos en 
listaA=[] #Lista de valores unicos de Case Number
listaB=[] #Lista de valores repetidos de Case Number
for i in range(1,len(df)+1):
    if df["Case Number"].loc[i] not in listaA:
        listaA.append(df["Case Number"].loc[i])
    else:
        listaB.append(df["Case Number"].loc[i])

In [21]:
#Vamos a ver que valores se repiten en Case Number.1 y si son los mismos en 
listaC=[] #Lista de valores unicos de Case Number.1
listaD=[] #Lista de valores repetidos de Case Number.1
for i in range(1,len(df)+1):
    if df["Case Number.1"].loc[i] not in listaC:
        listaC.append(df["Case Number.1"].loc[i])
    else:
        listaD.append(df["Case Number.1"].loc[i])

In [22]:
#Vamos a ver que valores se repiten en Case Number.2 y si son los mismos en 
listaE=[] #Lista de valores unicos de Case Number.2
listaF=[] #Lista de valores repetidos de Case Number.2
for i in range(1,len(df)+1):
    if df["Case Number.2"].loc[i] not in listaE:
        listaE.append(df["Case Number.2"].loc[i])
    else:
        listaF.append(df["Case Number.2"].loc[i])

In [23]:
#Vamos a comprobar que columnas son mas parecidas
print((df["Case Number"]==df["Case Number.1"]).value_counts())
print((df["Case Number.1"]==df["Case Number.2"]).value_counts())
print((df["Case Number"]==df["Case Number.2"]).value_counts())

True     5979
False      13
dtype: int64
True     5981
False      11
dtype: int64
True     5990
False       2
dtype: int64


In [24]:
#Primero unificare la columna Case Number con la columna Case Number.2 ya que son muy parecidas
print(f'Valores no repetidos {listaA==listaE}')
print(f'Valores  repetidos {listaB==listaF}')


Valores no repetidos False
Valores  repetidos True


In [25]:
#Los dos valores que no coinciden estan en los valores no repetidos
for i in range(1,len(df)):
    if df["Case Number"].loc[i]!= df["Case Number.2"].loc[i]:
          print(i)
    


1273
5823


In [26]:
#Analizamos ambos valores de forma individual
df.loc[1273]

Case Number                                                    1934.01.08.R
Date                                                   Reported 08-Feb-1934
Year                                                                   1934
Type                                                                Boating
Country                                                              TURKEY
Area                                                               Istanbul
Location                                         Haydarpasa jetty, Istanbul
Activity                                                            Fishing
Name                                                                2 males
Sex                                                                       M
Age                                                                     NaN
Injury                                                            No injury
Fatal (Y/N)                                                               N
Time        

In [27]:
df.loc[5823]

Case Number                                                      2015.07-10
Date                                                              10-Jul-15
Year                                                                   2015
Type                                                             Unprovoked
Country                                                                 USA
Area                                                             California
Location                                    Huntington Beach, Orange County
Activity                                                            Surfing
Name                                                           Danny Miskin
Sex                                                                       M
Age                                                                      38
Injury                              No injury, shark bumped & damaged board
Fatal (Y/N)                                                               N
Time        

- Problema:  
     - Case Number: A veces apuntdo por fecha o por el nombre del pdf, valores repetidos y si son el mismo dia con letras, no da informacion relevante ni utilizable ya que son parches de otras columnas

Decido borrar las columnas Case number ya que no me aportan informacion relevante y lo he unido a la columna Casos

#### Output Columnas: "original order","Case Number","Case Number.1","Case Number.2"

In [28]:
df.drop(["Case Number","Case Number.1","Case Number.2"],axis=1,inplace=True)

In [29]:
df

Date  Year        Type             Country               Area  \
Casos                                                                         
1        1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   Eastern Province   
2        1883-1889     0  Unprovoked              PANAMA                NaN   
3        1900-1905     0  Unprovoked                 USA     North Carolina   
4      Before 1903     0  Unprovoked           AUSTRALIA  Western Australia   
5      Before 1903     0  Unprovoked           AUSTRALIA  Western Australia   
...            ...   ...         ...                 ...                ...   
5988     16-Sep-16  2016  Unprovoked           AUSTRALIA           Victoria   
5989     17-Sep-16  2016  Unprovoked           AUSTRALIA           Victoria   
5990     18-Sep-16  2016  Unprovoked                 USA            Florida   
5991     18-Sep-16  2016  Unprovoked                 USA            Florida   
5992     18-Sep-16  2016  Unprovoked                 USA            Florida   

                                  Location      Activity  \
Casos                                                      
1      Below the English fort, Trincomalee      Swimming   
2                     Panama Bay 8ºN, 79ºW           NaN   
3                           Ocracoke Inlet      Swimming   
4                                      NaN  Pearl diving   
5                              Roebuck Bay        Diving   
...                                    ...           ...   
5988                           Bells Beach       Surfing   
5989                      Thirteenth Beach       Surfing   
5990      New Smyrna Beach, Volusia County       Surfing   
5991      New Smyrna Beach, Volusia County       Surfing   
5992      New Smyrna Beach, Volusia County       Surfing   

                        Name Sex   Age  \
Casos                                    
1                       male    M   15   
2            Jules Patterson    M  NaN   
3      Coast Guard personnel    M  NaN   
4                      Ahmun    M  NaN   
5                       male    M  NaN   
...                      ...  ...  ...   
5988                    male    M  NaN   
5989         Rory Angiolella    M  NaN   
5990                    male    M   43   
5991          Chucky Luciano    M   36   
5992                    male    M   16   

                                                  Injury Fatal (Y/N)   Time  \
Casos                                                                         
1      FATAL. "Shark bit him in half, carrying away t...           Y    NaN   
2                                                  FATAL           Y    NaN   
3                                                  FATAL           Y    NaN   
4                                                  FATAL           Y    NaN   
5                                                  FATAL           Y    NaN   
...                                                  ...         ...    ...   
5988               No injury: Knocked off board by shark           N    NaN   
5989                        Struck by fin on chest & leg           N    NaN   
5990                            Lacerations to lower leg           N  10h43   
5991                                Lacerations to hands           N  11h00   
5992                               Minor injury to thigh           N  13h00   

        Species                  Investigator or Source  \
Casos                                                     
1            NaN                             S.W. Baker   
2            NaN                    The Sun, 10/20/1938   
3            NaN   F. Schwartz, p.23; C. Creswell, GSAF   
4            NaN  H. Taunton; N. Bartlett,  pp. 233-234   
5            NaN       H. Taunton; N. Bartlett,  p. 234   
...          ...                                    ...   
5988   2 m shark                     The Age, 9/16/2016   
5989         NaN                     The Age, 9/18/2016   
5990         NaN            Orlando Sent

### Columna "unname 23"

####   analisis de los valores y conocimiento del problema

In [30]:
#Valores de Unnamed 23
set(df["Unnamed: 23"])

{'Teramo', 'change filename', nan}

In [31]:
#Cantidad de veces que salen estos valores
df["Unnamed: 23"].value_counts()

change filename    1
Teramo             1
Name: Unnamed: 23, dtype: int64

#### Conclusion y output

In [32]:
# No la considero relevante asi que decido borrarla
df.drop("Unnamed: 23",axis=1,inplace=True)

### Columna "unname 22"

#### Identificacion de valores nulos de cada columna

In [33]:
#Valores de Unnamed 22
set(df["Unnamed: 22"])

{nan, 'stopped here'}

In [34]:
#Vamos a ver cuantos valores hay
df["Unnamed: 22"].value_counts()

stopped here    1
Name: Unnamed: 22, dtype: int64

#### Conclusion y output

In [35]:
#El registro esta en mitad de la tabla y no lo considero importante asi que quito toda la columna 
df.drop("Unnamed: 22", axis=1,inplace=True)

###  Columnas: "Pais"  

#### Analisis inicial de las columnas

In [134]:
#Observamos la cantidad de valores nulos que hay
(df["Country"]).isna().sum()

0

In [133]:
#Rellenamos los na con un grupo de "No determinado"
df["Country"].fillna("No determinado",inplace=True)

####  Limpieza de los valores:
Observo que en los inicios hay espacios, hay ?, primero limpiar los nombres y despues buscar si existe el pais
Objetivo: comprobar si existen y si no existen añadir como No determinado, antes borrar espacios iniciales y finales y quitar ?¿ç

In [222]:
#Primero lo convierto todo a mayusculas
df["Country"]=df["Country"].str.upper()


In [223]:
#Identifico los paises que empiezan con espacio y se lo quito
df[df["Country"].str.startswith(" ")]["Country"]
#Solo son PHILIPPINES Y TONGA, los modifico 
df.replace({" PHILIPPINES":"PHILIPPINES"," TONGA":"TONGA"},inplace=True)

In [206]:
#Identifico los paises que acaban con espacio y se lo quito
df[df["Country"].str.endswith(" ")]["Country"]
df.replace({"YEMEN ":"YEMEN",
            "EGYPT ":"EGYPT",
           "NORTH ATLANTIC OCEAN ":"NORTH ATLANTIC OCEAN",
           "NICARAGUA ":"NICARAGUA",
            "PACIFIC OCEAN ":"PACIFIC OCEAN",
            "MEXICO ":"MEXICO"
           },inplace=True)

In [204]:
#Identifico los valores que tienen ? y los modifico
for i in df["Country"]:
    if re.findall("[?]",i):
        print(i)

YEMEN 
YEMEN 
YEMEN 
YEMEN 
YEMEN 
EGYPT 
EGYPT 
YEMEN 
YEMEN 
NORTH ATLANTIC OCEAN 
NICARAGUA 
PACIFIC OCEAN 
PACIFIC OCEAN 
MEXICO 


In [194]:
df.replace({"ASIA?":"ASIA",
            "RED SEA?":"RED SEA",
            "INDIAN OCEAN?":"INDIAN OCEAN",
            "MEDITERRANEAN SEA?":"MEDITERRANEAN SEA",
           "SUDAN?":"SUDAN"},inplace=True)

####  Conclusion: Clasificacion por paises existentes
los nombres de los paises ya estan limpios, vamos a comprobar si existen 
[Github del csv](https://github.com/datasets/country-list/blob/master/data.csv#L119)

In [215]:
#Descargo de github un csv con todos los paises del mundo en ingles, viene con indice y codigo asi que creo una lista con los paises
lista_paises=pd.read_csv("paises.csv")
lista_paises=list(lista_paises["Name"])

In [225]:
#Convierto los nombres a mayuscula y los pongo en una lista
lista=[]
for i in lista_paises:
    lista.append(i.upper())
lista_paises=lista

In [226]:
lista_paises

['AFGHANISTAN',
 'ÅLAND ISLANDS',
 'ALBANIA',
 'ALGERIA',
 'AMERICAN SAMOA',
 'ANDORRA',
 'ANGOLA',
 'ANGUILLA',
 'ANTARCTICA',
 'ANTIGUA AND BARBUDA',
 'ARGENTINA',
 'ARMENIA',
 'ARUBA',
 'AUSTRALIA',
 'AUSTRIA',
 'AZERBAIJAN',
 'BAHAMAS',
 'BAHRAIN',
 'BANGLADESH',
 'BARBADOS',
 'BELARUS',
 'BELGIUM',
 'BELIZE',
 'BENIN',
 'BERMUDA',
 'BHUTAN',
 'BOLIVIA',
 'BONAIRE',
 'BOSNIA AND HERZEGOVINA',
 'BOTSWANA',
 'BOUVET ISLAND',
 'BRAZIL',
 'BRITISH INDIAN OCEAN TERRITORY',
 'BRUNEI DARUSSALAM',
 'BULGARIA',
 'BURKINA FASO',
 'BURUNDI',
 'CAMBODIA',
 'CAMEROON',
 'CANADA',
 'CAPE VERDE',
 'CAYMAN ISLANDS',
 'CENTRAL AFRICAN REPUBLIC',
 'CHAD',
 'CHILE',
 'CHINA',
 'CHRISTMAS ISLAND',
 'COCOS (KEELING) ISLANDS',
 'COLOMBIA',
 'COMOROS',
 'CONGO',
 'DEMOCRATIC REPUBLIC OF THE CONGO',
 'COOK ISLANDS',
 'COSTA RICA',
 "CÔTE D'IVOIRE",
 'CROATIA',
 'CUBA',
 'CURAÇAO',
 'CYPRUS',
 'CZECH REPUBLIC',
 'DENMARK',
 'DJIBOUTI',
 'DOMINICA',
 'DOMINICAN REPUBLIC',
 'ECUADOR',
 'EGYPT',
 'EL SALVADOR

In [208]:
#Vamos a ver los paises que no identifica para ver cuales podemos solucionar y cuales no
for i in set(df["Country"]):
    if i in list(country_list):
        continue
    else:
        print(i)

CEYLON (SRI LANKA)
MID-PACIFC OCEAN
ASIA
PERSIAN GULF
SAN DOMINGO
NEW BRITAIN
TRINIDAD & TOBAGO
NORTH SEA
UNITED ARAB EMIRATES (UAE)
SOUTH PACIFIC OCEAN
FALKLAND ISLANDS
SOUTHWEST PACIFIC OCEAN
OKINAWA
EQUATORIAL GUINEA / CAMEROON
BRITISH ISLES
NETHERLANDS ANTILLES
GULF OF ADEN
BURMA
BRITISH NEW GUINEA
THE BALKANS
FEDERATED STATES OF MICRONESIA
PACIFIC OCEAN
MEDITERRANEAN SEA
MALDIVE ISLANDS
ANDAMAN / NICOBAR ISLANDAS
TASMAN SEA
REUNION
BAY OF BENGAL
BAHREIN
ATLANTIC OCEAN
DIEGO GARCIA
TURKS & CAICOS
PALESTINIAN TERRITORIES
SYRIA
NO DETERMINADO
JAVA
RUSSIA
GRAND CAYMAN
NEVIS
ST. MAARTIN
JOHNSTON ISLAND
ST. MARTIN
NORTH PACIFIC OCEAN
ENGLAND
ST HELENA
ANTIGUA
NORTH ATLANTIC OCEAN
BRITISH WEST INDIES
AZORES
IRAN / IRAQ
SCOTLAND
SOUTH ATLANTIC OCEAN
SOLOMON ISLANDS / VANUATU
SOUTH CHINA SEA
COLUMBIA
RED SEA
BETWEEN PORTUGAL & INDIA
ADMIRALTY ISLANDS
WESTERN SAMOA
CENTRAL PACIFIC
OCEAN
EGYPT / ISRAEL
COAST OF AFRICA
CRETE
ITALY / CROATIA
KOREA
VIETNAM
BRITISH VIRGIN ISLANDS
NORTHERN ARABIA

**MODIFICACIONES sobre la lista de paises que no aparecen** {Como salen: Transformacion} (ND=no determinado)
1. CEYLON (SRI LANKA): SRI LANKA
1. MID-PACIFC OCEAN:ND
1. ASIA:ND
1.  PERSIAN GULF: ND
1.  SAN DOMINGO: DOMINICAN REPUBLIC
1.  NEW BRITAIN: ND
1.  TRINIDAD & TOBAGO: TRINIDAD AND TOBAGO
1.  NORTH SEA: ND
1.  UNITED ARAB EMIRATES (UAE): 'UNITED ARAB EMIRATES'
1.  SOUTH PACIFIC OCEAN: ND
1.  FALKLAND ISLANDS: 'FALKLAND ISLANDS (MALVINAS)'
1.  SOUTHWEST PACIFIC OCEAN: ND
1.  OKINAWA: JAPAN
1.  EQUATORIAL GUINEA / CAMEROON: ND
1.  BRITISH ISLES: ND
1.  NETHERLANDS ANTILLES: ND
1.  GULF OF ADEN: YEMEN
1.  BURMA: MYANMAR
1.  BRITISH NEW GUINEA: ND
1.  THE BALKANS: ND
1.  FEDERATED STATES OF MICRONESIA: MICRONESIA
1.  PACIFIC OCEAN: ND
1.  MEDITERRANEAN SEA: ND
1.  MALDIVE ISLANDS: MALDIVES
1.  ANDAMAN / NICOBAR ISLANDAS: INDIA
1.  TASMAN SEA: AUSTRALIA
1.  REUNION: FRANCE
1.  BAY OF BENGAL:INDIA
1.  BAHREIN: BAHRAIN
1.  ATLANTIC OCEAN: ND
1.  DIEGO GARCIA: ND
1.  TURKS & CAICOS: ND PERO MUY BONITO
1.  PALESTINIAN TERRITORIES: PALESTINE
1.  SYRIA: 'SYRIAN ARAB REPUBLIC'
1.  NO DETERMINADO
1.  JAVA: INDONESIA
1.  RUSSIA: RUSSIAN FEDERATION
1.  GRAND CAYMAN: CAYMAN ISLANDS'
1.  NEVIS: ND
1.  ST. MAARTIN: 'SAINT MARTIN (FRENCH PART)'
1.  JOHNSTON ISLAND: ND
1.  ST. MARTIN: 'SAINT MARTIN (FRENCH PART)'
1.  NORTH PACIFIC OCEAN: ND
1.  ENGLAND:'UNITED KINGDOM'
1.  ST HELENA: 'SAINT HELENA'
1.  ANTIGUA: 'ANTIGUA AND BARBUDA'
1.  NORTH ATLANTIC OCEAN: ND
1.  BRITISH WEST INDIES: ND
1.  AZORES: PORTUGAL
1.  IRAN / IRAQ: ND
1.  SCOTLAND: UNITED KINGDOM
1.  SOUTH ATLANTIC OCEAN: ND
1.  SOLOMON ISLANDS / VANUATU: ND
1.  SOUTH CHINA SEA: CHINA
1.  COLUMBIA: COLOMBIA
1.  RED SEA: ND
1.  BETWEEN PORTUGAL & INDIA: ND
1.  ADMIRALTY ISLANDS: ND
1.  WESTERN SAMOA: SAMOA
1.  CENTRAL PACIFIC: ND
1.  OCEAN: ND
1.  EGYPT / ISRAEL: ND
1.  COAST OF AFRICA: ND
1.  CRETE: GREECE
1.  ITALY / CROATIA: ND
1.  KOREA: ND
1.  VIETNAM: VIET NAM
1.  BRITISH VIRGIN ISLANDS: ND
1.  NORTHERN ARABIAN SEA: ND
1.  INDIAN OCEAN: ND
1.  USA: 'UNITED STATES'
1.  CARIBBEAN SEA: ND
1.  RED SEA / INDIAN OCEAN: ND
1.  NEW GUINEA: GUINEA
1.  CURACAO: BRAZIL
1.  MID ATLANTIC OCEAN: ND

In [233]:
#Hacemos las modificaciones en el dataset
df.replace({"CEYLON (SRI LANKA)": "SRI LANKA",
            "SAN DOMINGO": "DOMINICAN REPUBLIC",
            "TRINIDAD & TOBAGO": "TRINIDAD AND TOBAGO",
            "UNITED ARAB EMIRATES (UAE)": "UNITED ARAB EMIRATES",
           "FALKLAND ISLANDS": "FALKLAND ISLANDS (MALVINAS)",
            "OKINAWA": "JAPAN",
           "GULF OF ADEN": "YEMEN",
            "BURMA": "MYANMAR",
           "FEDERATED STATES OF MICRONESIA": "MICRONESIA",
            "MALDIVE ISLANDS": "MALDIVES",
           "ANDAMAN / NICOBAR ISLANDAS": "INDIA",
            "TASMAN SEA": "AUSTRALIA",
            "REUNION":"FRANCE",
            "BAY OF BENGAL":"INDIA",
            "BAHREIN":"BAHRAIN",
           "PALESTINIAN TERRITORIES":"PALESTINE",
            "SYRIA": 'SYRIAN ARAB REPUBLIC',
            "JAVA": "INDONESIA",
            "RUSSIA": "RUSSIAN FEDERATION",
            "GRAND CAYMAN": "CAYMAN ISLANDS",
            "ST. MAARTIN": 'SAINT MARTIN (FRENCH PART)',
            "ST. MARTIN": 'SAINT MARTIN (FRENCH PART)',
            "ENGLAND":'UNITED KINGDOM',
            "ST HELENA": 'SAINT HELENA',
            "ANTIGUA": 'ANTIGUA AND BARBUDA',
              "AZORES": "PORTUGAL",
            "SCOTLAND": "UNITED KINGDOM",
            "SOUTH CHINA SEA": "CHINA",
            "COLUMBIA": "COLOMBIA",
            "WESTERN SAMOA": "SAMOA",
            "CRETE": "GREECE",
            "VIETNAM": "VIET NAM",
            "USA": 'UNITED STATES',
            "NEW GUINEA": "GUINEA",
            "CURACAO": "BRAZIL"
            
           },inplace=True)

In [235]:
#Los paises que no estan bien escritos los añadimos a No determinado
for i in set(df["Country"]):
    if i in list(country_list):
        continue
    else:
        df.replace({i:"No determinado"},inplace=True)

In [238]:
#Comprobamos
df["Country"].value_counts()

UNITED STATES           2116
AUSTRALIA               1280
SOUTH AFRICA             565
No determinado           174
PAPUA NEW GUINEA         133
                        ... 
SYRIAN ARAB REPUBLIC       1
CYPRUS                     1
COOK ISLANDS               1
GABON                      1
PUERTO RICO                1
Name: Country, Length: 128, dtype: int64

### Columna "Age"

#### Identificacion de valores nulos de cada columna

In [36]:
df["Age"].isna().sum()

2681

#### Conclusion y output


In [37]:
#La mitad de los valores son na por lo que elimino la columna
df.drop("Age",axis=1,inplace=True)

###  Columnas: "pdf","href formula","href"

Objetivo unificar las 3 columnas al pdf de esa columna

####  Analisis inicial

In [38]:
#Observamos las 3 columnas
df[["pdf","href formula","href"]]

pdf  \
Casos                                   
1                  ND-0001-Ceylon.pdf   
2          ND-0002-JulesPatterson.pdf   
3      ND-0003-Ocracoke_1900-1905.pdf   
4                   ND-0004-Ahmun.pdf   
5              ND-0005-RoebuckBay.pdf   
...                               ...   
5988        2016.09.16-BellsBeach.pdf   
5989        2016.09.17-Angiolella.pdf   
5990             2016.09.18.a-NSB.pdf   
5991         2016.09.18.b-Luciano.pdf   
5992             2016.09.18.c-NSB.pdf   

                                            href formula  \
Casos                                                      
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
5      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
5988   http://sharkattackfile.net/spreadsheets/pdf_di...   
5989   http://sharkattackfile.net/spreadsheets/pdf_di...   
5990   http://sharkattackfile.net/spreadsheets/pdf_di...   
5991   http://sharkattackfile.net/spreadsheets/pdf_di...   
5992   http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                    href  
Casos                                                     
1      http://sharkattackfile.net/spreadsheets/pdf_di...  
2      http://sharkattackfile.net/spreadsheets/pdf_di...  
3      http://sharkattackfile.net/spreadsheets/pdf_di...  
4      http://sharkattackfile.net/spreadsheets/pdf_di...  
5      http://sharkattackfile.net/spreadsheets/pdf_di...  
...                                                  ...  
5988   http://sharkattackfile.net/spreadsheets/pdf_di...  
5989   http://sharkattackfile.net/spreadsheets/pdf_di...  
5990   http://sharkattackfile.net/spreadsheets/pdf_di...  
5991   http://sharkattackfile.net/spreadsheets/pdf_di...  
5992   http://sharkattackfile.net/spreadsheets/pdf_di...  

[5992 rows x 3 columns]

In [39]:
#Vamos a comprobar los valores nulos de cada columna para elegir la que menos tenga
df[["pdf","href formula","href"]].isna().sum()

pdf             0
href formula    1
href            3
dtype: int64

In [40]:
df[["pdf","href formula","href"]].isna().sort_values("href formula")


pdf  href formula   href
Casos                            
1      False         False  False
4000   False         False  False
3999   False         False  False
3998   False         False  False
3997   False         False  False
...      ...           ...    ...
1993   False         False  False
1992   False         False  False
1991   False         False  False
2000   False         False  False
2973   False          True  False

[5992 rows x 3 columns]

In [41]:
df.loc[3020,["pdf","href formula","href"]]

pdf                                      1976.00.00.b-Nakaima.pdf
href formula    http://sharkattackfile.net/spreadsheets/pdf_di...
href            http://sharkattackfile.net/spreadsheets/pdf_di...
Name: 3020, dtype: object

#### Conclusion y output

In [42]:
#añado la formula de href a la columna con todos los hipervinculos
df["href formula"].at[3020]=df["href"].at[3020]

In [43]:
#elimino las columnas pdf y href 
df.drop(["pdf","href"],axis=1,inplace=True)

In [100]:
#cambio de nombre a la columna que guardo en el df
df.rename(columns={"href formula":"Pdf"},inplace=True)

###  Columnas: "species"

####   Analisis inicial

In [45]:
#Comprobamos cuantos valores na hay
df["Species "].isna().sum()

2934

In [46]:
# Cuantas categorias de especies diferentes hay
df["Species "].nunique()

1538

In [47]:
set(df["Species "])

{nan,
 "6 m [20'] white shark",
 "Reported to involve a White shark, 6 m to 7m [20' to 23'] ",
 '.5 m shark',
 "Raggedtooth shark, 1.5 m [5'] (tooth fragment recovered)",
 "2.3 m [7.5'] white shark, identified by M. Smale",
 '\xa0 ',
 'Said to involve a 1 m  white shark, but thought  that it was more likely a blue shark',
 "Tiger shark, 3 to 4 m [10' to 13'] ",
 "Bronze whaler shark,3.7 m [12'] ",
 '9-foot shark',
 "White shark, 10 '",
 '"a black-tipped shark"',
 'Tiger shark',
 '1.5 m, 45-kg shark',
 '2.27 m shark',
 "1.5 to 2.5 m [5' to 8.25'] copper shark",
 "Mako shark, 5'",
 'Reported to involve a bull shark',
 "7' to 8' bull shark",
 "9' shark",
 "7' silvertip shark",
 'White sharks',
 '7-gill shark?',
 'Blue shark bites present ',
 '1.2 m to 1.5 m [4\' to 5\'] "reef shark"',
 "3 m [10'], 200-kg [441-lb] bull shark",
 '136-kg [300-lb] shark',
 "4 m [13'] shark",
 "1.7 m [5.5'] shark",
 'Shark involvement prior to death could not be determined',
 '4.2 m white shark',
 "6' to 8' bu

#### Conclusion y output

Mas de la mitad de los registros son desconocidos y de los que hay valores no NA, hay demasiada variacion en la forma de escribirlo y en la informacion que proporcionan, por lo que decido borrar la columna

In [48]:
df.drop("Species ",axis=1,inplace=True)

In [49]:
df

Date  Year        Type             Country               Area  \
Casos                                                                         
1        1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   Eastern Province   
2        1883-1889     0  Unprovoked              PANAMA                NaN   
3        1900-1905     0  Unprovoked                 USA     North Carolina   
4      Before 1903     0  Unprovoked           AUSTRALIA  Western Australia   
5      Before 1903     0  Unprovoked           AUSTRALIA  Western Australia   
...            ...   ...         ...                 ...                ...   
5988     16-Sep-16  2016  Unprovoked           AUSTRALIA           Victoria   
5989     17-Sep-16  2016  Unprovoked           AUSTRALIA           Victoria   
5990     18-Sep-16  2016  Unprovoked                 USA            Florida   
5991     18-Sep-16  2016  Unprovoked                 USA            Florida   
5992     18-Sep-16  2016  Unprovoked                 USA            Florida   

                                  Location      Activity  \
Casos                                                      
1      Below the English fort, Trincomalee      Swimming   
2                     Panama Bay 8ºN, 79ºW           NaN   
3                           Ocracoke Inlet      Swimming   
4                                      NaN  Pearl diving   
5                              Roebuck Bay        Diving   
...                                    ...           ...   
5988                           Bells Beach       Surfing   
5989                      Thirteenth Beach       Surfing   
5990      New Smyrna Beach, Volusia County       Surfing   
5991      New Smyrna Beach, Volusia County       Surfing   
5992      New Smyrna Beach, Volusia County       Surfing   

                        Name Sex   \
Casos                               
1                       male    M   
2            Jules Patterson    M   
3      Coast Guard personnel    M   
4                      Ahmun    M   
5                       male    M   
...                      ...  ...   
5988                    male    M   
5989         Rory Angiolella    M   
5990                    male    M   
5991          Chucky Luciano    M   
5992                    male    M   

                                                  Injury Fatal (Y/N)   Time  \
Casos                                                                         
1      FATAL. "Shark bit him in half, carrying away t...           Y    NaN   
2                                                  FATAL           Y    NaN   
3                                                  FATAL           Y    NaN   
4                                                  FATAL           Y    NaN   
5                                                  FATAL           Y    NaN   
...                                                  ...         ...    ...   
5988               No injury: Knocked off board by shark           N    NaN   
5989                        Struck by fin on chest & leg           N    NaN   
5990                            Lacerations to lower leg           N  10h43   
5991                                Lacerations to hands           N  11h00   
5992                               Minor injury to thigh           N  13h00   

                      Investigator or Source  \
Casos                                          
1                                 S.W. Baker   
2                        The Sun, 10/20/1938   
3       F. Schwartz, p.23; C. Creswell, GSAF   
4      H. Taunton; N. Bartlett,  pp. 233-234   
5           H. Taunton; N. Bartlett,  p. 234   
...                                      ...   
5988                      The Age, 9/16/2016   
5989                      The Age, 9/18/2016   
5990             Orlando Sentinel, 9/19/2016   
5991             Orlando Sentinel, 9/19/2016   
5992             Orlando Sentinel, 9/19/2016   

                                            href formula  
Casos      

### Columnas: "Date","Year","Time"

#### Analisis de las columnas, cosas en comun y diferencias

In [50]:
# Vamos con las columnas de Date,year y time
#Objetivo unificar a una sola tabla
df[["Date","Year", "Time"]].head(5)

Date  Year Time
Casos                        
1        1845-1853     0  NaN
2        1883-1889     0  NaN
3        1900-1905     0  NaN
4      Before 1903     0  NaN
5      Before 1903     0  NaN

In [51]:
df[["Date","Year", "Time"]].isna().sum()

Date       0
Year       0
Time    3213
dtype: int64

#### Agrupacion de las tres columnas en una

In [52]:
#Con tantos valor nulo en Time, la borramos
df.drop("Time",axis=1,inplace=True)

In [53]:
#Analizamos la variable Date
df["Date"].nunique()

5128

In [54]:
#la columna Date es tipo objeto con demasiados valores unicos, decido agruparla en la columna Año ya que no hay suficientes registros con los dias y los meses
df.Date.sample(n=9)

Casos
1796           1952
735          Jul-05
4337      03-Mar-02
520     21-Jul-1889
5177      15-Feb-10
1747      25-Jun-50
4608      05-Mar-05
2428      30-Sep-62
4486      14-Oct-03
Name: Date, dtype: object

In [55]:
#Esta columna ya son valores int, el problema son los casos ocurridos en el 0
df.Year.value_counts()

2015    139
2011    128
2014    125
0       124
2013    122
       ... 
1854      1
1850      1
1846      1
1834      1
1554      1
Name: Year, Length: 232, dtype: int64

**La cantidad de valores nulos y de valores diferentes de las columna Time y Date no lo veo aprovechable, me quedo con los años**

In [56]:
#Opciones con los años 0: Borrar esos registros, poner la media o la mediana
mediana=df.Year.median()
#elijo la mediana ya que creo que es mas correcto

In [57]:
#Creo una funcion para que me cambie los valores 0 por la mediana de los años
def años(año):
    if año==0:
        return int(mediana)
    else:
        return año


In [58]:
#aplico la funcion que cambiara los valores 0 al valor de la mediana
df["Year"]=df["Year"].apply(años)

In [59]:
df.Year.value_counts()

1975    173
2015    139
2011    128
2014    125
2013    122
       ... 
1850      1
1846      1
1834      1
1822      1
1554      1
Name: Year, Length: 231, dtype: int64

In [60]:
#elimino la columna Date
df.drop("Date",axis=1,inplace=True)

### Columnas: "Name" y "Sex "

#### Analisis inicial de las columnas

In [61]:
#Primero le quito el espacion al nombre de la columna "Sex "
df.columns=(['Year', 'Type', 'Country', 'Area', 'Location', 'Activity', 'Name',
       'Sex', 'Injury', 'Fatal (Y/N)', 'Investigator or Source',
       'href formula'])

In [62]:
df[["Sex","Name"]].isna().sum()

Sex     567
Name    200
dtype: int64

In [63]:
#Analisis de la columna name
df["Name"].nunique()

5009

In [64]:
#Demasiada variedad en el formato y en las anotaciones, otra columna que borramos y agruparemos por sexo
df["Name"].sample(n=8)

Casos
857                                                  male
1797                                     Kiaploki, a male
4575                                         Megan Durham
5277                                        Metereti Jeke
1347    22' boat, occupants: Saul White & Charles Dill...
5776                                         Rick Neumann
2302                                     James K. Stewart
4192                                       Margaret White
Name: Name, dtype: object

In [65]:
df[df["Sex"].isna()]

Year          Type          Country               Area  \
Casos                                                           
15     1975    Unprovoked             FIJI       Moala Island   
24     1975    Unprovoked          REUNION        Saint-Denis   
43     1975  Sea Disaster              NaN                NaN   
65     1975    Unprovoked        SINGAPORE                NaN   
83     1975    Unprovoked  SOLOMON ISLANDS   Malaita Province   
...     ...           ...              ...                ...   
5767   2015          Boat   ATLANTIC OCEAN                NaN   
5883   2015          Boat        AUSTRALIA    New South Wales   
5911   2016       Invalid              USA         California   
5963   2016          Boat              USA            Alabama   
5987   2016          Boat        AUSTRALIA  Western Australia   

                                                Location  \
Casos                                                      
15                                                   NaN   
24                                             Barachois   
43                                                   NaN   
65     Keppel Harbor,  2 miles from Singapore city ce...   
83                                       Waibana Passage   
...                                                  ...   
5767                                                 NaN   
5883                                      Lake Macquarie   
5911               Bolsa Chica State Park, Orange County   
5963                                  8 miles off Mobile   
5987                                             Bunbury   

                                         Activity  \
Casos                                               
15           Wreck of  large double sailing canoe   
24                                       Swimming   
43     A group of survivors on a raft for 17-days   
65                                       Swimming   
83                                         Diving   
...                                           ...   
5767                         Transatlantic Rowing   
5883                                      Fishing   
5911                                      Surfing   
5963        Fishing in Alabama Deep Fishing Rodeo   
5987                                      Fishing   

                                                Name  Sex  \
Casos                                                       
15                                        20 Fijians  NaN   
24                                               NaN  NaN   
43                                                C.  NaN   
65                                               NaN  NaN   
83                                      Albert Raiti  NaN   
...                                              ...  ...   
5767   Avalon, a carbon kevlar monohull: 8 occupants  NaN   
5883     6 m boat: occupants  Stephen & Andrew Crust  NaN   
5911                                         unknown  NaN   
5963                            Occupant: Ben Raines  NaN   
5987                          Occupant: Ben Stratton  NaN   

                                                  Injury Fatal (Y/N)  \
Casos                                                                  
15     FATAL, 18 people  were killed by sharks, 2 sur...           Y   
24                                                 FATAL           Y   
43     FATAL, shark leapt into raft and bit the man w...           Y   
65                                             Recovered           N   
83                         Lacerations to hands and knee           N   
...                                                  ...         ...   
5767                         No injury, shark bit rudder           N   
5883            No injury, shark rammed boat & bit motor           N   
5911         Board reportedly bumped by shark. No injury           N   
5963                 No injury, shark bit trolling motor           N   
5987            Shark rammed boa

In [66]:
#Vamos a analizar los valores
set(df["Sex"])

{'.', 'F', 'M', 'M ', 'N', 'lli', nan}

In [67]:
df["Sex"].value_counts()
#Homogenizamos los M, y el resto 

M      4835
F       585
M         2
.         1
N         1
lli       1
Name: Sex, dtype: int64

####  Conclusiones y cambios

In [68]:
#Los M con espacio los agrupo en "M", los valores "lli","." y "N" asi como los valores nulos iran a ND(No determinado)
def genero(x):
    if x=="M ":
        return "M"
    elif x=="M" or x=="F":
        return x
    else:
        return "ND"
    


In [69]:
df["Sex"]=df["Sex"].apply(genero)

In [70]:
#Borramos la columna name
df.drop("Name",axis=1,inplace=True)

### Columnas: "Injury" y "Fatal (Y/N) "

#### Analisis inicial de las columnas

In [71]:
df[["Injury","Fatal (Y/N)"]].isna().sum()

Injury         27
Fatal (Y/N)    19
dtype: int64

In [72]:
#cantidad de valores tiene cada uno
df[["Injury","Fatal (Y/N)"]].nunique()

Injury         3595
Fatal (Y/N)       8
dtype: int64

In [73]:
#No veo relevante el tantas categorias en injury
df[["Injury","Fatal (Y/N)"]].isna().sum()

Injury         27
Fatal (Y/N)    19
dtype: int64

####  Conclusiones y cambios

In [74]:
# No encuentro patron por lo que eliminaremos esta columna
df["Injury"].sample(n=5)

Casos
2006                 Injured by shark
4486                      Foot bitten
2863                   Bitten on legs
5279    Severe injury to left forearm
251                             FATAL
Name: Injury, dtype: object

In [75]:
#vamos a ver los valores de fatal
df["Fatal (Y/N)"].value_counts()

N          4315
Y          1552
UNKNOWN      94
 N            8
F             1
#VALUE!       1
N             1
n             1
Name: Fatal (Y/N), dtype: int64

Clasificacion de los valores:
- N: No es fatal, sobrevive
- Y: Fatal, No sobrevive
- UNKNOWN: Desconocido
- #VALUE!: Desconocido
- F: Fatal, no sobrevive
Creo una funcion para modificar los valores

In [76]:
#Vamos a cambiar los valores a: Si, No, Desconocido
def fatal(x):
    if x=="N" or x=="N " or x==" N" or x=="n":
        return "Sobrevive"
    elif x=="Y" or x=="F":
        return "No sobrevive"
    else:
        return "Desconocido"

In [77]:
#Aplicamos la funcion y cambio el nombre de la columna
df["Supervivencia"]=df["Fatal (Y/N)"].apply(fatal)

In [78]:
#Borramos las columnas de injury y fatal
df.drop(["Injury","Fatal (Y/N)"],axis=1,inplace=True)

### Columnas: "Activity" 

#### Analisis inicial de las columnas

In [79]:
df.Activity.nunique()

1492

####  Conclusiones y cambios

In [80]:
#Elijo las 14 mas comunes y el resto y los NA los clasificare como Otra actividad
lista=df["Activity"].value_counts().head(14)

In [81]:
#Funcion para cambiar los valores
def actividad(x):
    if x in lista:
        return x
    else:
        return "Otra actividad"

In [82]:
#Cambio los valores
df["Activity"]=df["Activity"].apply(actividad)

### Columnas:  "Type"

#### Analisis inicial de las columnas

In [83]:
#No hay valores nulos
df.Type.isna().sum()

0

In [84]:
df["Type"].nunique()

6

In [85]:
df["Type"].value_counts()

Unprovoked      4386
Provoked         557
Invalid          519
Sea Disaster     220
Boat             200
Boating          110
Name: Type, dtype: int64

####  Conclusiones y cambios

In [86]:
#Junto los valores de Boat y boating en Boat
def typeBoat(x):
    if x=="Boating":
        return "Boat"
    else:
        return x

In [87]:
df["Type"]=df["Type"].apply(typeBoat)

In [88]:
df.Type.value_counts()

Unprovoked      4386
Provoked         557
Invalid          519
Boat             310
Sea Disaster     220
Name: Type, dtype: int64

### Columna: "Investigator or Source"  

#### Analisis inicial de las columnas

In [90]:
#Hay muy pocos valores na
df["Investigator or Source"].isna().sum()

15

In [91]:
#Hay demasiados valores unicos
df["Investigator or Source"].nunique()

4752

In [92]:
df["Investigator or Source"].sample(n=10)

Casos
1756    Evening Sun (Baltimore), 8/11/1950; Long Beach...
2720                                (1994) SAF Case #1524
5550                                  ABC News, 6/18/2013
174                       Times of London, 8/5/1822, p. 2
4885                                            wsbtv.com
2886                            Canberra Times, 6/26/1972
1668     A. M. Rapson, p.147; D. Baldridge, pp.128 & 261 
824                          Poverty Bay Herald, 3/9/1910
2156                            Times of India, 9/10/1959
785                         Dawson Daily News, 11/20/1907
Name: Investigator or Source, dtype: object

Por lo que veo aparece el autor o la revista, la pagina y la fecha pero con diferentes formatos
- Opcion de rellenar los na como desconocido

In [244]:
df


Year        Type        Country               Area  \
Casos                                                       
1      1975  Unprovoked      SRI LANKA   Eastern Province   
2      1975  Unprovoked         PANAMA                NaN   
3      1975  Unprovoked  UNITED STATES     North Carolina   
4      1975  Unprovoked      AUSTRALIA  Western Australia   
5      1975  Unprovoked      AUSTRALIA  Western Australia   
...     ...         ...            ...                ...   
5988   2016  Unprovoked      AUSTRALIA           Victoria   
5989   2016  Unprovoked      AUSTRALIA           Victoria   
5990   2016  Unprovoked  UNITED STATES            Florida   
5991   2016  Unprovoked  UNITED STATES            Florida   
5992   2016  Unprovoked  UNITED STATES            Florida   

                                  Location        Activity Sex  \
Casos                                                            
1      Below the English fort, Trincomalee        Swimming   M   
2                     Panama Bay 8ºN, 79ºW  Otra actividad   M   
3                           Ocracoke Inlet        Swimming   M   
4                                      NaN  Otra actividad   M   
5                              Roebuck Bay          Diving   M   
...                                    ...             ...  ..   
5988                           Bells Beach         Surfing   M   
5989                      Thirteenth Beach         Surfing   M   
5990      New Smyrna Beach, Volusia County         Surfing   M   
5991      New Smyrna Beach, Volusia County         Surfing   M   
5992      New Smyrna Beach, Volusia County         Surfing   M   

                      Investigator or Source  \
Casos                                          
1                                 S.W. Baker   
2                        The Sun, 10/20/1938   
3       F. Schwartz, p.23; C. Creswell, GSAF   
4      H. Taunton; N. Bartlett,  pp. 233-234   
5           H. Taunton; N. Bartlett,  p. 234   
...                                      ...   
5988                      The Age, 9/16/2016   
5989                      The Age, 9/18/2016   
5990             Orlando Sentinel, 9/19/2016   
5991             Orlando Sentinel, 9/19/2016   
5992             Orlando Sentinel, 9/19/2016   

                                                     Pdf Supervivencia  
Casos                                                                   
1      http://sharkattackfile.net/spreadsheets/pdf_di...  No sobrevive  
2      http://sharkattackfile.net/spreadsheets/pdf_di...  No sobrevive  
3      http://sharkattackfile.net/spreadsheets/pdf_di...  No sobrevive  
4      http://sharkattackfile.net/spreadsheets/pdf_di...  No sobrevive  
5      http://sharkattackfile.net/spreadsheets/pdf_di...  No sobrevive  
...                                                  ...           ...  
5988   http://sharkattackfile.net/spreadsheets/pdf_di...     Sobrevive  
5989   http://sharkattackfile.net/spreadsheets/pdf_di...     Sobrevive  
5990   http://sharkattackfile.net/spreadsheets/pdf_di...     Sobrevive  
5991   http://sharkattackfile.net/spreadsheets/pdf_di...     Sobrevive  
5992   http://sharkattackfile.net/spreadsheets/pdf_di...     Sobrevive  

[5992 rows x 10 columns]

In [155]:
df["Country"].apply(pais).value_counts()

No especificado             2475
AUSTRALIA                   1279
SOUTH AFRICA                 565
PAPUA NEW GUINEA             133
NEW ZEALAND                  125
                            ... 
SLOVENIA                       1
NORTHERN MARIANA ISLANDS       1
SWEDEN                         1
MONACO                         1
ALGERIA                        1
Name: Country, Length: 114, dtype: int64

In [199]:
"SPAIN" in country_list

False

In [157]:
for i in country_list:
    if i=="CEYLON":
        print("SI")
   

In [212]:
country_list[country_list.str.startswith("B")]

16                           BAHAMAS
17                           BAHRAIN
18                        BANGLADESH
19                          BARBADOS
20                           BELARUS
21                           BELGIUM
22                            BELIZE
23                             BENIN
24                           BERMUDA
25                            BHUTAN
26                           BOLIVIA
27                           BONAIRE
28            BOSNIA AND HERZEGOVINA
29                          BOTSWANA
30                     BOUVET ISLAND
31                            BRAZIL
32    BRITISH INDIAN OCEAN TERRITORY
33                 BRUNEI DARUSSALAM
34                          BULGARIA
35                      BURKINA FASO
36                           BURUNDI
Name: Name, dtype: object

### Columnas: "Pais"  

#### Analisis inicial de las columnas

In [134]:
#Observamos la cantidad de valores nulos que hay
(df["Country"]).isna().sum()

0

In [133]:
#Rellenamos los na con un grupo de "No determinado"
df["Country"].fillna("No determinado",inplace=True)

####  Limpieza de los valores:
Observo que en los inicios hay espacios, hay ?, primero limpiar los nombres y despues buscar si existe el pais
Objetivo: comprobar si existen y si no existen añadir como No determinado, antes borrar espacios iniciales y finales y quitar ?¿ç

In [222]:
#Primero lo convierto todo a mayusculas
df["Country"]=df["Country"].str.upper()


In [223]:
#Identifico los paises que empiezan con espacio y se lo quito
df[df["Country"].str.startswith(" ")]["Country"]
#Solo son PHILIPPINES Y TONGA, los modifico 
df.replace({" PHILIPPINES":"PHILIPPINES"," TONGA":"TONGA"},inplace=True)

In [206]:
#Identifico los paises que acaban con espacio y se lo quito
df[df["Country"].str.endswith(" ")]["Country"]
df.replace({"YEMEN ":"YEMEN",
            "EGYPT ":"EGYPT",
           "NORTH ATLANTIC OCEAN ":"NORTH ATLANTIC OCEAN",
           "NICARAGUA ":"NICARAGUA",
            "PACIFIC OCEAN ":"PACIFIC OCEAN",
            "MEXICO ":"MEXICO"
           },inplace=True)

In [204]:
#Identifico los valores que tienen ? y los modifico
for i in df["Country"]:
    if re.findall("[?]",i):
        print(i)

YEMEN 
YEMEN 
YEMEN 
YEMEN 
YEMEN 
EGYPT 
EGYPT 
YEMEN 
YEMEN 
NORTH ATLANTIC OCEAN 
NICARAGUA 
PACIFIC OCEAN 
PACIFIC OCEAN 
MEXICO 


In [194]:
df.replace({"ASIA?":"ASIA",
            "RED SEA?":"RED SEA",
            "INDIAN OCEAN?":"INDIAN OCEAN",
            "MEDITERRANEAN SEA?":"MEDITERRANEAN SEA",
           "SUDAN?":"SUDAN"},inplace=True)

####  Existen como paises los valores?
los nombres de los paises ya estan limpios, vamos a comprobar si existen 
[Github del csv](https://github.com/datasets/country-list/blob/master/data.csv#L119)

In [215]:
#Descargo de github un csv con todos los paises del mundo en ingles, viene con indice y codigo asi que creo una lista con los paises
lista_paises=pd.read_csv("paises.csv")
lista_paises=list(lista_paises["Name"])

In [225]:
#Convierto los nombres a mayuscula y los pongo en una lista
lista=[]
for i in lista_paises:
    lista.append(i.upper())
lista_paises=lista

In [226]:
lista_paises

['AFGHANISTAN',
 'ÅLAND ISLANDS',
 'ALBANIA',
 'ALGERIA',
 'AMERICAN SAMOA',
 'ANDORRA',
 'ANGOLA',
 'ANGUILLA',
 'ANTARCTICA',
 'ANTIGUA AND BARBUDA',
 'ARGENTINA',
 'ARMENIA',
 'ARUBA',
 'AUSTRALIA',
 'AUSTRIA',
 'AZERBAIJAN',
 'BAHAMAS',
 'BAHRAIN',
 'BANGLADESH',
 'BARBADOS',
 'BELARUS',
 'BELGIUM',
 'BELIZE',
 'BENIN',
 'BERMUDA',
 'BHUTAN',
 'BOLIVIA',
 'BONAIRE',
 'BOSNIA AND HERZEGOVINA',
 'BOTSWANA',
 'BOUVET ISLAND',
 'BRAZIL',
 'BRITISH INDIAN OCEAN TERRITORY',
 'BRUNEI DARUSSALAM',
 'BULGARIA',
 'BURKINA FASO',
 'BURUNDI',
 'CAMBODIA',
 'CAMEROON',
 'CANADA',
 'CAPE VERDE',
 'CAYMAN ISLANDS',
 'CENTRAL AFRICAN REPUBLIC',
 'CHAD',
 'CHILE',
 'CHINA',
 'CHRISTMAS ISLAND',
 'COCOS (KEELING) ISLANDS',
 'COLOMBIA',
 'COMOROS',
 'CONGO',
 'DEMOCRATIC REPUBLIC OF THE CONGO',
 'COOK ISLANDS',
 'COSTA RICA',
 "CÔTE D'IVOIRE",
 'CROATIA',
 'CUBA',
 'CURAÇAO',
 'CYPRUS',
 'CZECH REPUBLIC',
 'DENMARK',
 'DJIBOUTI',
 'DOMINICA',
 'DOMINICAN REPUBLIC',
 'ECUADOR',
 'EGYPT',
 'EL SALVADOR

In [208]:
#Vamos a ver los paises que no identifica para ver cuales podemos solucionar y cuales no
for i in set(df["Country"]):
    if i in list(country_list):
        continue
    else:
        print(i)

CEYLON (SRI LANKA)
MID-PACIFC OCEAN
ASIA
PERSIAN GULF
SAN DOMINGO
NEW BRITAIN
TRINIDAD & TOBAGO
NORTH SEA
UNITED ARAB EMIRATES (UAE)
SOUTH PACIFIC OCEAN
FALKLAND ISLANDS
SOUTHWEST PACIFIC OCEAN
OKINAWA
EQUATORIAL GUINEA / CAMEROON
BRITISH ISLES
NETHERLANDS ANTILLES
GULF OF ADEN
BURMA
BRITISH NEW GUINEA
THE BALKANS
FEDERATED STATES OF MICRONESIA
PACIFIC OCEAN
MEDITERRANEAN SEA
MALDIVE ISLANDS
ANDAMAN / NICOBAR ISLANDAS
TASMAN SEA
REUNION
BAY OF BENGAL
BAHREIN
ATLANTIC OCEAN
DIEGO GARCIA
TURKS & CAICOS
PALESTINIAN TERRITORIES
SYRIA
NO DETERMINADO
JAVA
RUSSIA
GRAND CAYMAN
NEVIS
ST. MAARTIN
JOHNSTON ISLAND
ST. MARTIN
NORTH PACIFIC OCEAN
ENGLAND
ST HELENA
ANTIGUA
NORTH ATLANTIC OCEAN
BRITISH WEST INDIES
AZORES
IRAN / IRAQ
SCOTLAND
SOUTH ATLANTIC OCEAN
SOLOMON ISLANDS / VANUATU
SOUTH CHINA SEA
COLUMBIA
RED SEA
BETWEEN PORTUGAL & INDIA
ADMIRALTY ISLANDS
WESTERN SAMOA
CENTRAL PACIFIC
OCEAN
EGYPT / ISRAEL
COAST OF AFRICA
CRETE
ITALY / CROATIA
KOREA
VIETNAM
BRITISH VIRGIN ISLANDS
NORTHERN ARABIA

**MODIFICACIONES sobre la lista de paises que no aparecen** {Como salen: Transformacion} (ND=no determinado)
1. CEYLON (SRI LANKA): SRI LANKA
1. MID-PACIFC OCEAN:ND
1. ASIA:ND
1.  PERSIAN GULF: ND
1.  SAN DOMINGO: DOMINICAN REPUBLIC
1.  NEW BRITAIN: ND
1.  TRINIDAD & TOBAGO: TRINIDAD AND TOBAGO
1.  NORTH SEA: ND
1.  UNITED ARAB EMIRATES (UAE): 'UNITED ARAB EMIRATES'
1.  SOUTH PACIFIC OCEAN: ND
1.  FALKLAND ISLANDS: 'FALKLAND ISLANDS (MALVINAS)'
1.  SOUTHWEST PACIFIC OCEAN: ND
1.  OKINAWA: JAPAN
1.  EQUATORIAL GUINEA / CAMEROON: ND
1.  BRITISH ISLES: ND
1.  NETHERLANDS ANTILLES: ND
1.  GULF OF ADEN: YEMEN
1.  BURMA: MYANMAR
1.  BRITISH NEW GUINEA: ND
1.  THE BALKANS: ND
1.  FEDERATED STATES OF MICRONESIA: MICRONESIA
1.  PACIFIC OCEAN: ND
1.  MEDITERRANEAN SEA: ND
1.  MALDIVE ISLANDS: MALDIVES
1.  ANDAMAN / NICOBAR ISLANDAS: INDIA
1.  TASMAN SEA: AUSTRALIA
1.  REUNION: FRANCE
1.  BAY OF BENGAL:INDIA
1.  BAHREIN: BAHRAIN
1.  ATLANTIC OCEAN: ND
1.  DIEGO GARCIA: ND
1.  TURKS & CAICOS: ND PERO MUY BONITO
1.  PALESTINIAN TERRITORIES: PALESTINE
1.  SYRIA: 'SYRIAN ARAB REPUBLIC'
1.  NO DETERMINADO
1.  JAVA: INDONESIA
1.  RUSSIA: RUSSIAN FEDERATION
1.  GRAND CAYMAN: CAYMAN ISLANDS'
1.  NEVIS: ND
1.  ST. MAARTIN: 'SAINT MARTIN (FRENCH PART)'
1.  JOHNSTON ISLAND: ND
1.  ST. MARTIN: 'SAINT MARTIN (FRENCH PART)'
1.  NORTH PACIFIC OCEAN: ND
1.  ENGLAND:'UNITED KINGDOM'
1.  ST HELENA: 'SAINT HELENA'
1.  ANTIGUA: 'ANTIGUA AND BARBUDA'
1.  NORTH ATLANTIC OCEAN: ND
1.  BRITISH WEST INDIES: ND
1.  AZORES: PORTUGAL
1.  IRAN / IRAQ: ND
1.  SCOTLAND: UNITED KINGDOM
1.  SOUTH ATLANTIC OCEAN: ND
1.  SOLOMON ISLANDS / VANUATU: ND
1.  SOUTH CHINA SEA: CHINA
1.  COLUMBIA: COLOMBIA
1.  RED SEA: ND
1.  BETWEEN PORTUGAL & INDIA: ND
1.  ADMIRALTY ISLANDS: ND
1.  WESTERN SAMOA: SAMOA
1.  CENTRAL PACIFIC: ND
1.  OCEAN: ND
1.  EGYPT / ISRAEL: ND
1.  COAST OF AFRICA: ND
1.  CRETE: GREECE
1.  ITALY / CROATIA: ND
1.  KOREA: ND
1.  VIETNAM: VIET NAM
1.  BRITISH VIRGIN ISLANDS: ND
1.  NORTHERN ARABIAN SEA: ND
1.  INDIAN OCEAN: ND
1.  USA: 'UNITED STATES'
1.  CARIBBEAN SEA: ND
1.  RED SEA / INDIAN OCEAN: ND
1.  NEW GUINEA: GUINEA
1.  CURACAO: BRAZIL
1.  MID ATLANTIC OCEAN: ND

In [233]:
#Hacemos las modificaciones en el dataset
df.replace({"CEYLON (SRI LANKA)": "SRI LANKA",
            "SAN DOMINGO": "DOMINICAN REPUBLIC",
            "TRINIDAD & TOBAGO": "TRINIDAD AND TOBAGO",
            "UNITED ARAB EMIRATES (UAE)": "UNITED ARAB EMIRATES",
           "FALKLAND ISLANDS": "FALKLAND ISLANDS (MALVINAS)",
            "OKINAWA": "JAPAN",
           "GULF OF ADEN": "YEMEN",
            "BURMA": "MYANMAR",
           "FEDERATED STATES OF MICRONESIA": "MICRONESIA",
            "MALDIVE ISLANDS": "MALDIVES",
           "ANDAMAN / NICOBAR ISLANDAS": "INDIA",
            "TASMAN SEA": "AUSTRALIA",
            "REUNION":"FRANCE",
            "BAY OF BENGAL":"INDIA",
            "BAHREIN":"BAHRAIN",
           "PALESTINIAN TERRITORIES":"PALESTINE",
            "SYRIA": 'SYRIAN ARAB REPUBLIC',
            "JAVA": "INDONESIA",
            "RUSSIA": "RUSSIAN FEDERATION",
            "GRAND CAYMAN": "CAYMAN ISLANDS",
            "ST. MAARTIN": 'SAINT MARTIN (FRENCH PART)',
            "ST. MARTIN": 'SAINT MARTIN (FRENCH PART)',
            "ENGLAND":'UNITED KINGDOM',
            "ST HELENA": 'SAINT HELENA',
            "ANTIGUA": 'ANTIGUA AND BARBUDA',
              "AZORES": "PORTUGAL",
            "SCOTLAND": "UNITED KINGDOM",
            "SOUTH CHINA SEA": "CHINA",
            "COLUMBIA": "COLOMBIA",
            "WESTERN SAMOA": "SAMOA",
            "CRETE": "GREECE",
            "VIETNAM": "VIET NAM",
            "USA": 'UNITED STATES',
            "NEW GUINEA": "GUINEA",
            "CURACAO": "BRAZIL"
            
           },inplace=True)

In [235]:
#Los paises que no estan bien escritos los añadimos a No determinado
for i in set(df["Country"]):
    if i in list(country_list):
        continue
    else:
        df.replace({i:"No determinado"},inplace=True)

In [238]:
#Comprobamos
df["Country"].value_counts()

UNITED STATES           2116
AUSTRALIA               1280
SOUTH AFRICA             565
No determinado           174
PAPUA NEW GUINEA         133
                        ... 
SYRIAN ARAB REPUBLIC       1
CYPRUS                     1
COOK ISLANDS               1
GABON                      1
PUERTO RICO                1
Name: Country, Length: 128, dtype: int64